### PR TITLE
Liquidation edge cases

### PR DESCRIPTION
**Pull request by mts1715**
Closes: #173 

**_Note_**: should be reviewed after merging https://github.com/onflow/FlowALP/pull/172
Some of proposed test in task weren't implemented because Flow doesn't have gas-price-based ordering at all.

## Summary

Adds mainnet fork tests for liquidation edge cases: partial sequences, multi-collateral seizure, DEX liquidity failures, oracle-deviation circuit breaker, fee accrual over time, and bad debt handling.

## Tests added (fork_liquidation_edge_cases.cdc)

1. `testPartialLiquidationSequences`

Five positions with distinct collateral types (FLOW, USDF, USDC, WETH, WBTC), each borrowing MOET at health ≈ 1.1. After a FLOW price crash, only the FLOW position becomes unhealthy (health 0.95). A liquidator partially restores it across three sequential calls (seize 10 / repay 20 MOET each), stepping health through 0.967 → 0.985 → 1.005. A fourth call and a second liquidator's attempt both revert once the position is healthy. Verifies that positions backed by unaffected collateral are untouched.

2. `testLiquidateMultiCollateralChooseUSDC`

A single position holds FLOW + USDC + WETH as collateral with USDF debt (health ≈ 1.109). After a FLOW crash ($1.00 → $0.75), the position drops to health ≈ 0.935. The liquidator selectively seizes USDC (seize 40 USDC, repay 55 USDF) while FLOW and WETH balances remain untouched. Validates that a liquidator can choose the optimal collateral without disturbing other assets.

3. `testDexLiquidityConstraints`

A MockDexSwapper vault is seeded with only 50% of the required repayment tokens (23 USDF instead of 46). The batch DEX liquidation reverts atomically, leaving position state unchanged. After topping up the DEX vault to 53 USDF, the identical parameters succeed. Verifies that insufficient DEX liquidity causes a clean, state-preserving failure rather than a partial execution.

4. `testStabilityAndInsuranceFeeAccrual`

Sets a 10% annual fixed interest rate on USDF with 10% stability fee and 10% insurance fee. An LP deposits 5000 USDF; a borrower takes 130 USDF against 200 FLOW. After Test.moveTime(by: ONE_YEAR), debt compounds to ≈ 143.67 USDF (continuous compounding: 130 × e^0.10). A liquidator then partially restores health. Fee collection is verified:

Stability fund receives ≈ 0.705 USDF (debitBalance 67 × (e^0.10 − 1) × 10%)
Insurance fund receives ≈ 0.705 MOET (swapped 1:1 via MockDex)
LP earns ≈ 416.435 USDF credit income (5000 × (e^0.08 − 1) — FixedRate applies creditRate to the full LP deposit, not just outstanding debt)

## Tests added (liquidation_phase1_test.cdc‎) instead of `testBadDebtHandling`

1. `testBadDebt_seizeAllCollateral`
A severe price crash drives a position into bad-debt territory (collateral market value < debt).
A liquidator can seize all remaining collateral in one call, leaving a zombie position with health = 0.0 and non-zero debt. FlowALPv0 has no automatic write-off mechanism — the shortfall is an implicit loss absorbed by LP reserve dilution.

2. `testBadDebt_voluntaryRepayment_fullRepayReverts`
After a position reaches bad-debt state (health = 0.0, no collateral), a liquidator can voluntarily repay the remaining debt with seize=0. However, repaying the very last token (which would set debt=0) reverts: postHealth=UFix128.max exceeds the 1.05 liquidationTargetHF.

------
**UPD**:
Not implemeted
_This test seems to mostly duplicate the coverage area of the tests matching testManualLiquidation_dexOraclePriceDivergence.* in cadence/tests/liquidation_phase1_test.cdc._

 `testLiquidationSlippageConstraints`

Governance tightens dexOracleDeviationBps from the default 300 bps to 200 bps. Two manual liquidation attempts use the same seize/repay amounts but different DEX price ratios:

Scenario 1 (priceRatio 0.7275, deviation ≈ 309 bps > 200 bps max) → reverts, position unchanged.
Scenario 2 (priceRatio 0.7425, deviation ≈ 101 bps < 200 bps max) → succeeds, post-health ≈ 1.036.
Validates the oracle-deviation guard that prevents liquidators from extracting value at stale DEX prices.

### CODE REVIEW

File: cadence/tests/fork_liquidation_edge_cases.cdc [hunk 1]
```diff
@@ -0,0 +1,967 @@
...
+
+    let detailsAfterLiq = getPositionDetails(pid: pid, beFailed: false)
+    // 200 - 55 = 145
+    Test.assertEqual(145.0, getCreditBalanceForType(details: detailsAfterLiq, vaultType: CompositeType(MAINNET_FLOW_TOKEN_ID)!))
+    // 130 - 46 = 84
+    Test.assertEqual(84.0, getDebitBalanceForType(details: detailsAfterLiq, vaultType: CompositeType(MAINNET_USDF_TOKEN_ID)!))
+}
+
+// =============================================================================
+// Stability and Insurance Fee Accrual — 1 year before liquidation
```
Comment (jordanschalm) [inline cadence/tests/fork_liquidation_edge_cases.cdc]: What's the idea behind this test case? What kinds of interactions between liquidation and fee collection are we trying to validate?

File: cadence/tests/fork_liquidation_edge_cases.cdc [hunk 2]
```diff
@@ -0,0 +1,967 @@
...
+    Test.assert(postHealth > crashedHealth)
+
+    // Verify seizure: 55 FLOW seized, 46 USDF repaid
+    let detailsAfterLiq = getPositionDetails(pid: pid, beFailed: false)
+    Test.assertEqual(145.0, getCreditBalanceForType(details: detailsAfterLiq, vaultType: CompositeType(MAINNET_FLOW_TOKEN_ID)!)) // 200 - 55 = 145
+    Test.assertEqual(84.0,  getDebitBalanceForType(details: detailsAfterLiq, vaultType: CompositeType(MAINNET_USDF_TOKEN_ID)!))  // 130 - 46 = 84
+}
+
+// =============================================================================
+// Liquidation Slippage Constraints
```
Comment (jordanschalm) [inline cadence/tests/fork_liquidation_edge_cases.cdc]: This test seems to mostly duplicate the coverage area of the tests matching `testManualLiquidation_dexOraclePriceDivergence.*` in `cadence/tests/liquidation_phase1_test.cdc`.

File: cadence/tests/fork_liquidation_edge_cases.cdc [hunk 3]
```diff
@@ -0,0 +1,967 @@
...
+        Test.reset(to: snapshot)
+    }
+}
+
+access(all) fun setup() {
+    deployContracts()
+
+    createAndStorePool(signer: MAINNET_PROTOCOL_ACCOUNT, defaultTokenIdentifier: MAINNET_MOET_TOKEN_ID, beFailed: false)
+
+    // Setup pool with real mainnet token prices
```
[the remark this answers is not in the document]
Comment (mts1715) [reply cadence/tests/fork_liquidation_edge_cases.cdc]: fix https://github.com/onflow/FlowALP/pull/181/commits/e5ef699199a583f58165d6a3d5d278915a6d4d20
